datapath: Reduce from LXC complexity

aspsk opened the pull request:
[ upstream commit https://github.com/cilium/cilium/commit/7575ba03ccaa5038255c8c1e5c31a8011ed9aaa1 ]

[ Backport note: This is a partial "cherry pick". Due to conflicts it was
  cherry-picked manually, chunk-by-chunk. It is partial because it doesn't
  include original changes for IPv6).

  This change also includes the following minor fix: https://github.com/cilium/cilium/commit/8dbde7237c6c040a079d37358f19b6ed00d37e62 ("test/bpf: Fix
  format of check-complexity.sh script"). ]

See the commit message for the original description.

<!-- Description of change -->

Fixes: #25500

```release-note
Reduce complexity of bpf_lxc by splitting per-packet lb to its own tail call
```

aspsk committed:
datapath: Reduce from LXC complexity

[ upstream commit 7575ba03ccaa5038255c8c1e5c31a8011ed9aaa1 ]

[ Backport note: Due to many conflicts it was cherry-picked manually.

  This change also includes the following minor fix: 8dbde7237c ("test/bpf: Fix
  format of check-complexity.sh script"). ]

Split from LXC code paths into two tail calls when per packet load
balancing is needed. When per packet load balancing is not needed this
should have minimal impact on datapath performance.

Kernel 4.9 verifier was unhappy when service load balancing code was
removed from handle_ipv6_from_lxc() and replaced with simpler state
restoration call, but only if host firewall was enabled, and per-packet
LB was on. Had to shuffle code around to make verifier happy again.

Shuffled IPv4 code to keep them similar, but git diff looks scarier
there.

Finally had to conditionally apply a revalidate data call to make
verifier happy also in the verifier complexity tests (test-verifier).

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

YutaroHayakawa commented on bpf/lib/lb.h:
```diff
@@ -873,6 +873,40 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 	return DROP_NO_SERVICE;
 }
+
+/* lb6_ctx_store_state() stores per packet load balancing state to be picked
+ * up on the continuation tail call.
+ * Note that the IP headers are already xlated and the tuple is re-initialized
+ * from the xlated headers before restoring state.
+ * NOTE: if lb_skip_l4_dnat() this is not the case as xlate is skipped. We
+ * lose the updated tuple daddr in that case.
+ */
+static __always_inline void lb6_ctx_store_state(struct __ctx_buff *ctx,
+						const struct ct_state *state)
+{
+	ctx_store_meta(ctx, CB_BACKEND_ID, state->backend_id);
+	ctx_store_meta(ctx, CB_CT_STATE, (__u32)state->rev_nat_index);
+}
+
+/* lb6_ctx_restore_state() restores per packet load balancing state from the
+ * previous tail call.
+ * tuple->flags does not need to be restored, as it will be reinitialized from
+ * the packet.
+ */
+static __always_inline void lb6_ctx_restore_state(struct __ctx_buff *ctx,
+						  struct ct_state *state)
+{
+	state->rev_nat_index = (__u16)ctx_load_meta(ctx, CB_CT_STATE);
+	/* Clear to not leak state to later stages of the datapath. */
+	ctx_store_meta(ctx, CB_CT_STATE, 0);
+
+	/* No loopback support for IPv6, see lb6_local() above. */
+
+	state->backend_id = ctx_load_meta(ctx, CB_BACKEND_ID);
+	/* Must clear to avoid policy bypass as CB_BACKEND_ID aliases CB_POLICY. */
+	ctx_store_meta(ctx, CB_BACKEND_ID, 0);
+}
```
Nit: Do we need this?

aspsk replied:
Technically, no. This is needed if we want to backport the ipv6 part. I didn't backport the ipv6 part because it slowed down port twice, and is not strictly required (ipv6 tests seem to pass). But probably, this is a good follow up to actually backport it, so I would just leave this code here.

YutaroHayakawa reviewed (approved):
LGTM with a nit. Thanks for working on this! 🙏

aspsk committed:
test/upgrade: disable check for the number of long-term connections

When downgrading from this version to v1.10 (and v1.11.{0,1}) some long-term
connections may be reset which breaks the CI, e.g.,

    /home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:527
    migrate-svc restart count values do not match
    Expected
        <int>: 0
    to be identical to
        <int>: 5
    /home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-net-next/src/github.com/cilium/cilium/test/k8sT/Updates.go:347

Disable this check and document in the Upgrade Guide.

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

pchaigno commented on Documentation/operations/upgrade.rst:
```diff
@@ -316,6 +316,12 @@ Annotations:
+----------------------
+
+* If downgrading to Cilium version <= v1.11.1 this is possible that some
+  long-term connections be interrupted due to missed tail calls.
```
That sentence doesn't parse to me :sweat_smile: 

```
If downgrading to Cilium version <= v1.11.1, some connections may be interrupted due to missed tail calls.
```
